src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+                itr.read() >> entry;
+                break;
+            }
+        }
+
+        auto integrateEntryValue = entry.get("Desktop Entry/X-AppImage-Integrate", "true");
+
+        boost::to_lower(integrateEntryValue);
+        boost::algorithm::trim(integrateEntryValue);
+
+        return integrateEntryValue == "false";
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    return -1;
+}
+
+
+/*
+ * Checks whether an AppImage's desktop file has set Terminal=true.
+ * Useful to check whether the author of an AppImage doesn't want it to be integrated.
+ *
+ * Returns >0 if set, 0 if not set, <0 on errors.
+ */
+int appimage_is_terminal_app(const char* path) {
+    try {
+        AppImage appImage(path);
+
+        std::vector<char> data;
+
+        XdgUtils::DesktopEntry::DesktopEntry entry;
+        // Load Desktop Entry
+        for (auto itr = appImage.files(); itr != itr.end(); ++itr) {
+            const auto& entryPath = *itr;
+            if (entryPath.find(".desktop") != std::string::npos && entryPath.find("/") == std::string::npos) {
+                itr.read() >> entry;
+                break;
+            }
+        }
+
+        auto terminalEntryValue = entry.get("Desktop Entry/Terminal");
+
+        boost::to_lower(terminalEntryValue);
+        boost::algorithm::trim(terminalEntryValue);
+
+        return terminalEntryValue == "true";
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+    return -1;
+}
+
+
+/* Return the md5 hash constructed according to
+ * https://specifications.freedesktop.org/thumbnail-spec/thumbnail-spec-latest.html#THUMBSAVE
+ * This can be used to identify files that are related to a given AppImage at a given location */
+char* appimage_get_md5(const char* path) {
+    using namespace appimage::utils;
+
+    if (path == nullptr)
+        return nullptr;
+
+    try {
+        const auto& canonicalPath = bf::weakly_canonical(path);
+
+        if (canonicalPath.empty())
+            return nullptr;
+
+        std::string uri = "file://" + UrlEncoder::encode(canonicalPath.string());
+
+        auto md5raw = HashLib::md5(uri);
+        auto md5Str = HashLib::toHex(md5raw);
+
+        return strdup(md5Str.c_str());
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+    return nullptr;
+}
+
+#ifdef LIBAPPIMAGE_DESKTOP_INTEGRATION_ENABLED
+using namespace appimage::desktop_integration;
```

Comment:
Sometimes you insert an empty line between preprocessor statements and other code, sometimes you don't. Please revise all headers that contain `#ifdef`s, and establish a consistent code style.